src/request.rs

```diff
@@ -293,6 +300,55 @@ impl<State> Request<State> {
         })?;
         Ok(res)
     }
+
+    /// returns a `Cookie` by name of the cookie.
+    pub fn get_cookie(&self, name: &str) -> Result<Option<Cookie<'static>>, Error> {
+        let cookie_data = self
+            .local::<CookieData>()
+            .ok_or_else(|| StringError(MIDDLEWARE_MISSING_MSG.to_owned()))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        let locked_jar = cookie_data
+            .content
+            .read()
+            .map_err(|e| StringError(format!("Failed to get read lock: {}", e)))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        Ok(locked_jar.get(name).cloned())
+    }
+
+    /// Add cookie to the cookie jar.
+    pub fn set_cookie(&mut self, cookie: Cookie<'static>) -> Result<(), Error> {
+        let cookie_data = self
+            .local::<CookieData>()
+            .ok_or_else(|| StringError(MIDDLEWARE_MISSING_MSG.to_owned()))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+
+        let mut locked_jar = cookie_data
+            .content
+            .write()
+            .map_err(|e| StringError(format!("Failed to get read lock: {}", e)))
+            .with_err_status(StatusCode::INTERNAL_SERVER_ERROR)?;
+        locked_jar.add(cookie);
+        Ok(())
+    }
```

Comment:
Actually, I think it would make more sense if these methods lived on `Response` instead. Before when we still had `Context` there was essentially a single shared mutable state. But that's no longer true, and it's a bit odd we're setting / removing values on the `Request`.

The way this could be implemented is if `Response` would hold its own `CookieJar`, and when the middleware runs we extract all values from the request's `CookieJar` and add them to the `Response`'s `CookieJar` using [`set_original`](https://docs.rs/cookie/0.12.0/cookie/struct.CookieJar.html#method.add_original).

## Example
That way the end-user API would be:

```rust
async fn set_cookie(_req: Request<()>) -> Response {
    let mut res = tide::Response::new(200);
    res.set_cookie(Cookie::new("hello", "world")).unwrap();
    res
}

Reply:
never thought about it, but makes actually much more sense.

Reply:
working on a solution here. I think could be done but not with `CookieJar`. The problem we need to fill the jar before the user can interact with it. Otherwise if the uses removes a cookie, that was not there before the Jar won't recognize is. This means the Set-Cookie won't be sent with empty value and thus not removed.

A workaround could be to track actions in the response rather than manipulation a cookie jar and later replay in cookie middleware. Should actually work fine for "normal" user behaviour. In the end, I guess the overhead should be minimal considering a user will not set and remove cookies with the same name in the `Response`.

Reply:
@yoshuawuyts what do you think?